### PR TITLE
Github Actions Migration - move go-tests workflows to GHA

### PR DESCRIPTION
**Pull request by jmurret**
### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] not a security concern

### CODE REVIEW

File: .github/workflows/reusable-dev-build.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        # TODO(JM) - figure out why this fails and if it is needed
+      # # this normally runs as the first job and has nothing to attach; only used in main branch after rebuilding UI
+      # - uses: actions/download-artifact@v3
+      #   with:
+      #     name: root-path
+      #     path: .
+      - name: Build
+        run: make dev
+      # save dev build to pass to downstream jobs
+      - uses: actions/upload-artifact@master
```
Comment (loshz) [inline .github/workflows/reusable-dev-build.yml:32]:
```suggestion
      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # pin@v3.1.2
```

File: .github/workflows/reusable-unit-split.yml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          go-version-file: 'go.mod'
+      - name: Setup go mod cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
```
Comment (loshz) [inline .github/workflows/reusable-unit-split.yml:85]: You might be able to just do this? https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs
```suggestion
          go-version-file: 'go.mod'
          cache: true
```

File: .github/workflows/reusable-unit-split.yml [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        run: |
+          wget https://github.com/gotestyourself/gotestsum/releases/download/v${{env.GOTESTSUM_VERSION}}/gotestsum_${{env.GOTESTSUM_VERSION}}_linux_amd64.tar.gz
+          sudo tar -C /usr/local/bin -xzf gotestsum_${{env.GOTESTSUM_VERSION}}_linux_amd64.tar.gz
+          rm gotestsum_${{env.GOTESTSUM_VERSION}}_linux_amd64.tar.gz
+      - run: mkdir -p ${{env.TEST_RESULTS}}
+      - name: go mod download
+        working-directory: ${{inputs.directory}}
+        run: go mod download
+      - name: Download consul
+        uses: actions/download-artifact@v3
```
Comment (loshz) [inline .github/workflows/reusable-unit-split.yml:96]:
```suggestion
        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # pin@v3.0.2
```

File: .github/workflows/reusable-unit-split.yml [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+          --debug \
+          --rerun-fails=3 \
+          --rerun-fails-max-failures=40 \
+          --rerun-fails-report=/tmp/gotestsum-rerun-fails \
+          --packages="$PACKAGE_NAMES" \
+          --junitfile ${{env.TEST_RESULTS}}/gotestsum-report.xml -- \
+          -tags="${{env.GOTAGS}}" -p 2 \
+          ${GO_TEST_FLAGS-} \
+          -cover -coverprofile=coverage.txt
+      - uses: actions/upload-artifact@master
```
Comment (loshz) [inline .github/workflows/reusable-unit-split.yml:127]:
```suggestion
      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # pin@v3.1.2
```

File: .github/workflows/reusable-unit-split.yml [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        with:
+          name: test-results
+          path: ${{env.TEST_RESULTS}}
+      - uses: actions/upload-artifact@master
```
Comment (loshz) [inline .github/workflows/reusable-unit-split.yml:131]:
```suggestion
      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # pin@v3.1.2
```

File: .github/workflows/reusable-unit.yml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+              --debug \
+              --rerun-fails=3 \
+              --rerun-fails-max-failures=40 \
+              --rerun-fails-report=/tmp/gotestsum-rerun-fails \
+              --packages="$PACKAGE_NAMES" \
+              --junitfile ${{env.TEST_RESULTS}}/gotestsum-report.xml -- \
+              -tags="${{env.GOTAGS}}" \
+              ${GO_TEST_FLAGS-} \
+              -cover -coverprofile=coverage.txt
+      - uses: actions/upload-artifact@master
```
Comment (loshz) [inline .github/workflows/reusable-unit.yml:98]:
```suggestion
      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # pin@v3.1.2
```

File: .github/workflows/reusable-unit.yml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        run: |
+          wget https://github.com/gotestyourself/gotestsum/releases/download/v${{env.GOTESTSUM_VERSION}}/gotestsum_${{env.GOTESTSUM_VERSION}}_linux_amd64.tar.gz
+          sudo tar -C /usr/local/bin -xzf gotestsum_${{env.GOTESTSUM_VERSION}}_linux_amd64.tar.gz
+          rm gotestsum_${{env.GOTESTSUM_VERSION}}_linux_amd64.tar.gz
+      - run: mkdir -p ${{env.TEST_RESULTS}}
+      - name: go mod download
+        working-directory: ${{inputs.directory}}
+        run: go mod download
+      - name: Download consul
+        uses: actions/download-artifact@v3
```
Comment (loshz) [inline .github/workflows/reusable-unit.yml:66]:
```suggestion
        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # pin@v3.0.2
```

**Review by loshz — CHANGES_REQUESTED** [review]
Overall LGTM. Left a couple of version pinning changes.

2 comments/questions:
- You might be able to drop the manual caching actions and use the `cache: true` https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs
- Do we actually use `gotestsum` and the reports it generates?

**Comment by jmurret** [comment]
> Do we actually use gotestsum and the reports it generates?

We use gotestsum to run these by package and as far as the reports it generates, it will keep those as artifacts for 30 days.  So if we need to do smarter test splitting,namely ones that are based on timings, we can utilize the previous run's junit summary in the future.  At the moment, tho, it is not used unless someone wants to peruse it on a previous run.

File: .github/scripts/rerun_fails_report.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+set -eu -o pipefail
+
+report_filename="${1?report filename is required}"
+if [ ! -s "$report_filename" ]; then
+    echo "gotestsum rerun report file is empty or missing"
+    exit 0
+fi
+
+function report {
+    echo ":repeat: gotestsum re-ran some tests in https://github.com/hashicorp/consul/commit/$GITHUB_SHA/checks"
```
Comment (loshz) [inline .github/scripts/rerun_fails_report.sh:17]: Not critical, but maybe use the specific run id?
```suggestion
    echo ":repeat: gotestsum re-ran some tests in https://github.com/hashicorp/consul/actions/run/$GITHUB_RUN_ID"
```

File: .github/workflows/reusable-unit-split.yml [hunk 9]
```diff
@@ -0,0 +1,132 @@
+name: reusable-unit-split
+
+on:
+  workflow_call:
+    inputs:
+      directory:
+        required: true
+        type: string
+      go-version:
```
Comment (loshz) [inline .github/workflows/reusable-unit-split.yml:9]: We don't use this as we get the value from go.mod

File: .github/workflows/reusable-unit-split.yml [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+      uploaded-binary-name:
+        required: false
+        type: string
+        default: "consul-bin" 
+      runs-on:
+        description: An expression indicating which kind of runners to use.
+        required: false
+        type: string
+        default: '["custom", "linux", "xl"]'
+      args:
```
Comment (loshz) [inline .github/workflows/reusable-unit-split.yml:25]: Doesn't look like we use this?

File: .github/workflows/reusable-unit.yml [hunk 11]
```diff
@@ -0,0 +1,111 @@
+name: reusable-unit 
+
+on:
+  workflow_call:
+    inputs:
+      directory:
+        required: true
+        type: string
+      go-version:
```
Comment (loshz) [inline .github/workflows/reusable-unit.yml:9]: We don't use this.

File: .github/workflows/reusable-unit.yml [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+jobs:
+  go-test:
+    runs-on: ${{ fromJSON(inputs.runs-on) }}
+    steps:      
+      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # pin@v3.3.0
+      - uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # pin@v3.5.0
+        with:
+          go-version-file: 'go.mod'
+      - name: Setup go mod cache
+        uses: actions/cache@v3
```
Comment (loshz) [inline .github/workflows/reusable-unit.yml:48]:
```suggestion
        uses: actions/cache@69d9d449aced6a2ede0bc19182fadc3a0a42d2b0 # pin@v3.2.6
```

File: .github/workflows/reusable-unit.yml [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+              --packages="$PACKAGE_NAMES" \
+              --junitfile ${{env.TEST_RESULTS}}/gotestsum-report.xml -- \
+              -tags="${{env.GOTAGS}}" \
+              ${GO_TEST_FLAGS-} \
+              -cover -coverprofile=coverage.txt
+      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # pin@v3.1.2
+        with:
+          name: test-results
+          path: ${{env.TEST_RESULTS}}
+      - uses: actions/upload-artifact@master
```
Comment (loshz) [inline .github/workflows/reusable-unit.yml:102]:
```suggestion
      - uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce # pin@v3.1.2
```

File: .github/workflows/reusable-unit.yml (hunk 11 above)
Comment (jmurret) [reply .github/workflows/reusable-unit.yml]: removed.  thank you.

File: .github/workflows/reusable-unit-split.yml (hunk 9 above)
Comment (jmurret) [reply .github/workflows/reusable-unit-split.yml]: removed.

File: .github/workflows/build-distros.yml [hunk 14]
```diff
@@ -8,26 +8,34 @@ permissions:
...
 jobs:
-  check-go-mod:
-    runs-on: ubuntu-22.04
+  setup:
+    name: Setup
+    runs-on: ubuntu-latest
+    outputs:
+      compute-small: ${{ steps.setup-outputs.outputs.compute-small }}
+      compute-medium: ${{ steps.setup-outputs.outputs.compute-medium }}
+      compute-large: ${{ steps.setup-outputs.outputs.compute-medium }}
```
Comment (loshz) [inline .github/workflows/build-distros.yml:17]:
```suggestion
      compute-large: ${{ steps.setup-outputs.outputs.compute-large }}
```

File: .github/workflows/go-tests.yml [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+  GOTESTSUM_VERSION: 1.8.2
+
+jobs:
+  setup:
+    name: Setup
+    runs-on: ubuntu-latest
+    outputs:
+      compute-small: ${{ steps.setup-outputs.outputs.compute-small }}
+      compute-medium: ${{ steps.setup-outputs.outputs.compute-medium }}
+      compute-large: ${{ steps.setup-outputs.outputs.compute-medium }}
```
Comment (loshz) [inline .github/workflows/go-tests.yml:28]:
```suggestion
      compute-large: ${{ steps.setup-outputs.outputs.compute-large }}
```

**Review by loshz — APPROVED** [review]
👍

**Pull request merged**